File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int main() {
  int t;
  cin >> t;

  for (int i = 0; i < t; i++) {
    stack<int> stack;
    string str;
    cin >> str;
    for (int j = 0; j < str.length(); j++) {
      if (str[j] == '(')
        stack.push(1);
      else {
        if (stack.empty()) {
          stack.push(1);
          break;
        }
        stack.pop();
      }
    }
    if (stack.empty())
      cout << "YES"
           << "\n";
    else
      cout << "NO"
           << "\n";
  }
}

File: 백준/Silver/9012. 괄호/괄호.py
T = int(input())

stack = []

while T:
    ps = input()

    for p in ps:
        if p == '(':
            stack.append('(')
        elif p == ')':
            if len(stack):
                if stack[-1] == '(':
                    stack.pop()
                else:
                    stack.append(')')
            else:
                stack.append(')')

    if len(stack):
        print('NO')
    else:
        print('YES')
    
    stack.clear()
    T -= 1